newhelm/cache_helper.py

```diff
@@ -0,0 +1,87 @@
+# mypy: ignore-errors
+from abc import ABC, abstractmethod
+import json
+import os
+from pydantic import BaseModel
+from sqlitedict import SqliteDict  # type: ignore
+from typing import Any, Dict, Generic, Mapping, TypeVar, Optional, get_args
+
+from newhelm.sut import SUT
+from newhelm.typed_data import TypedData
+
+RequestType = TypeVar("RequestType", bound=BaseModel)
+ResponseType = TypeVar("ResponseType", bound=BaseModel)
+
+
+class CacheHelper(ABC, Generic[RequestType, ResponseType]):
+    @abstractmethod
+    def get_cached_response(self, request: RequestType) -> Optional[ResponseType]:
+        """Returns the SUT's cached response to a request."""
+
+    @abstractmethod
+    def update_cache(self, request: RequestType, response: ResponseType):
+        """Cache the SUT's response in-memory."""
+
+
+class SUTResponseCacheHelper(CacheHelper[RequestType, ResponseType]):
+    """When caching is enabled, use previously cached SUT responses.
+
+    When used, the local directory structure will look like this:
+    data_dir/
+        test_1/
+            cached_responses/
+                sut_1.jsonl
+                sut_2.jsonl
+        ...
+      ...
+    """
+
+    def __init__(
+        self,
+        data_dir: str,
+        sut: SUT,
+    ):
+        self.data_dir = data_dir
+        self.fname = f"{sut.__class__.__name__,}.sqlite"
+        sut_bases = get_args(sut.__orig_bases__[0])
+        self._type_request: RequestType = sut_bases[0]
+        self._type_response: ResponseType = sut_bases[1]
+        self.cached_responses: Mapping[str, TypedData] = self._load_cached_responses()
```

Comment:
For typing, this variable needs to be more specifically a `SqliteDict` to allow for the `commit` to work later. Same for the return type of `_load_cached_responses`